share/wake/lib/core/json.wake

```diff
@@ -154,18 +154,155 @@ export def x // y =
+        def normalizeOrder =
+            JObject _.sortJObjectKeys
+        structurallyEqual (normalizeOrder a) (normalizeOrder b)
+    _ _ =
+        structurallyEqual x y
+
+# Compare two JSON values for equality according to their digital representation.
+# Note that this most significantly compares `JObject`s according to the order
+# in which their keys are stored.
+export def structurallyEqual (x: JValue) (y: JValue): Boolean = match x y
   (JString  a) (JString  b) = a ==~ b
   (JInteger a) (JInteger b) = a ==  b
   (JDouble  a) (JDouble  b) = a ==. b
   (JBoolean a) (JBoolean b) = if a then b else !b
   JNull        JNull        = True
   (JObject  a) (JObject  b) =
     def helper (Pair (Pair k c) (Pair l d)) = k ==~ l && c ==/ d
-    if a.len != b.len then False else zip a b | forall helper
+    if a.len != b.len then
+      False
+    else
+      zip a b
+      | forall helper
   (JArray   a) (JArray   b) =
     def helper (Pair c d) = c ==/ d
     if a.len != b.len then False else zip a b | forall helper
   (JArray a) _ = exists (_ ==/ y) a
   _ (JArray b) = exists (_ ==/ x) b
   _ _ = False
+
+# Simplify a JSON value according to the rules given.
+# For example, Wake's implementation of JSON is slightly more lenient than the
+# base standards, so if maximum compatibility is desired in situations where
+# some other implementation may be parsing generated output,
+# `normalizeJSONBasic` may be used to target that least common denominator.
+export def normalizeJSON (fmt: JSONNormalize): JValue => Result JValue Error = match _
+    JString a =
+        getJSONNormalizeString fmt a
+        | rmap JString
+    JInteger a =
+        getJSONNormalizeInteger fmt a
+        | rmap JInteger
+    JDouble a =
+        getJSONNormalizeDouble fmt a
+        | rmap JDouble
+    JBoolean a =
+        getJSONNormalizeBoolean fmt a
+        | rmap JBoolean
+    JNull =
+        Pass JNull
+    JObject a =
+        getJSONNormalizeObject fmt a
+        | rmap JObject
+    JArray a =
+        getJSONNormalizeArray fmt a
+        | rmap JArray
+
+# The rules by which `normalizeJSON` will simplify JSON values.
+# If any particular function is unable to operate on every input value (e.g. a
+# `JObject` contains duplicate keys of types which cannot be combined), that
+# rule may return a `Fail` which halts the broader processing.
+export tuple JSONNormalize =
+    String: String => Result String Error
+    Integer: Integer => Result Integer Error
+    Double: Double => Result Double Error
+    Boolean: Boolean => Result Boolean Error
+    Object: List (Pair String JValue) => Result (List (Pair String JValue)) Error
+    Array: List JValue => Result (List JValue) Error
+
+# A basic normalization ruleset which returns inputs unchanged.
+export def normalizeJSONDefault: JSONNormalize =
+    JSONNormalize Pass Pass Pass Pass Pass Pass
+
+# Target the minimum JSON language described by the specifications.
+export def normalizeJSONBasic: JSONNormalize =
```

Comment:
Is normal json json4? If so, I'd include a '4' in the name and also add a 'json5' version, which does not mess around with Doubles.